qsize.c

```diff
@@ -95,8 +95,36 @@ struct qsize_opt_state {
 
 	size_t num_all_files, num_all_nonfiles, num_all_ignored;
 	uint64_t num_all_bytes;
+
+	struct inode_bucket_t {
+		size_t len, cap;
+		struct {
+			dev_t dev;
+			ino_t inode;
+		} *elem;
+	} inodes[256];
 };
 
+/* Returns whether the dev,inode pair is a new pair that hasn't been seen so far */
+static bool add_inode(struct qsize_opt_state *state, dev_t dev, ino_t inode) {
```

Comment:
although, given the average size of files in a pkg, and the simplicity of %s/%s to match in one go the dev/inode being unique or not, is probably not that bad

Reply:
I implemented a version using sprintf+libq/set as you mentioned. The performance of `qsize texlive-fontsextra` is:
- Without inodes patch: 60 milliseconds
- This patch: 68 ms
  - IPC=1.96 (instructions per clock according to `perf stat`)
- sprintf+libq/set: 320 ms
  - IPC=0.55, mainly due to stalled-cycles-backend
  - according to `perf record`, `add_set_unique` consumes 73% of time when qsize is compiled with -O2

Reply:
Hmm, I'm ok with the possible performance degradation for the gain of not having to maintain more code.  I think at macro-level the performance impact is hardly noticable.

Reply:
Ok, I will update the patch to use libq/set and update the commit comment.

In relation to this subject, in general it is extra work to develop and maintain an internal list/map/set implementation in C in any project that does not have a good reason to do so. C++ provides **type-safe** array-lists, linked-lists, and sorted or hashed maps&sets, and has a string type that automatically frees memory when the string variable goes, directly or indirectly, out of scope. Have you been thinking about converting some of the code from C to C++? For example, the first step in converting qsize.c to qsize.cc would most likely entail just adding extern "C" to header files in the project.

Reply:
I don't want to add C++ as requirement, besides the fact that it would exclude me as future contributor as well ;)

Reply:
I see. The entry cost to C++ for C programmers is lower when focusing on features that C++ has in common with C. For example, there is no need to use the `class` keyword in the code because C++ understands the `struct` keyword just as well as C, and the only difference between `class` and `struct` in C++ is the default access mode to the fields of the data type. An issue for ease of entry to C++ is that hash-sets and hash-maps have an unusual syntax for defining the hash function for user data types in the `std` namespace - but fortunately sorted sets/maps do not have such weird syntax. Sorted sets/maps can in certain cases perform slower than hashed ones, O(log(N)) vs O(1), but the sorted ones are easier as a learning startpoint.